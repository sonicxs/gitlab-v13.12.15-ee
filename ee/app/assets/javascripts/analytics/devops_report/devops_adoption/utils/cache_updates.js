import produce from 'immer';
import devopsAdoptionSegmentsQuery from '../graphql/queries/devops_adoption_segments.query.graphql';

export const addSegmentsToCache = (store, segments, variables) => {
  const sourceData = store.readQuery({
    query: devopsAdoptionSegmentsQuery,
    variables,
  });

  const data = produce(sourceData, (draftData) => {
    draftData.devopsAdoptionSegments.nodes = [
      ...draftData.devopsAdoptionSegments.nodes,
      ...segments,
    ];
  });

  store.writeQuery({
    query: devopsAdoptionSegmentsQuery,
    variables,
    data,
  });
};

export const deleteSegmentsFromCache = (store, segmentIds, variables) => {
  const sourceData = store.readQuery({
    query: devopsAdoptionSegmentsQuery,
    variables,
  });

  const updatedData = produce(sourceData, (draftData) => {
    draftData.devopsAdoptionSegments.nodes = draftData.devopsAdoptionSegments.nodes.filter(
      ({ id }) => !segmentIds.includes(id),
    );
  });

  store.writeQuery({
    query: devopsAdoptionSegmentsQuery,
    variables,
    data: updatedData,
  });
};
