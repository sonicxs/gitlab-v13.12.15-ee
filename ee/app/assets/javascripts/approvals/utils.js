import { flatten } from 'lodash';

export const joinRuleResponses = (responsesArray) =>
  Object.assign({}, ...responsesArray, {
    rules: flatten(responsesArray.map(({ rules }) => rules)),
  });
