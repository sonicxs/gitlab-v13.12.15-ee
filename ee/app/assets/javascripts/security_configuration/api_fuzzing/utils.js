/* eslint-disable @gitlab/require-i18n-strings */
import { isString } from 'lodash';

export const insertTip = ({ snippet, tip, token }) => {
  if (!isString(snippet)) {
    throw new Error('snippet must be a string');
  }
  if (!isString(tip)) {
    throw new Error('tip must be a string');
  }
  if (!isString(token)) {
    throw new Error('token must be a string');
  }
  const lines = snippet.split('\n');
  for (let i = 0; i < lines.length; i += 1) {
    if (lines[i].includes(token)) {
      const indent = lines[i].match(/^[ \t]+/)?.[0] ?? '';
      lines[i] = lines[i].replace(token, `# ${tip}\n${indent}${token}`);
      break;
    }
  }
  return lines.join('\n');
};

export const insertTips = (snippet, tips = []) =>
  tips.reduce(
    (snippetWithTips, { tip, token }) => insertTip({ snippet: snippetWithTips, tip, token }),
    snippet,
  );
