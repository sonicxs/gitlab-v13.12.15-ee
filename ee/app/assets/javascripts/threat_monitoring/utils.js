/**
 * Get the height of the wrapper page element
 * This height can be used to determine where the highest element goes in a page
 * Useful for gl-drawer's header-height prop
 * @param {String} class the content wrapper class
 * @returns {String} height in px
 */
export const getContentWrapperHeight = (contentWrapperClass) => {
  const wrapperEl = document.querySelector(contentWrapperClass);
  return wrapperEl ? `${wrapperEl.offsetTop}px` : '';
};
