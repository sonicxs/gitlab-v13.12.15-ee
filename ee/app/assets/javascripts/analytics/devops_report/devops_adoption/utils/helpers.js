import { isToday } from '~/lib/utils/datetime_utility';

/**
 * A helper function which accepts the segments,
 *
 * @param {Object} params the segment data, timestamp and check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ segments, timestamp, openModal }) => {
  if (openModal) {
    return false;
  } else if (!segments.length) {
    return true;
  }

  const anyPendingSegments = segments.some((node) => node.latestSnapshot === null);
  const dataNotRefreshedToday = !isToday(new Date(timestamp));

  return anyPendingSegments || dataNotRefreshedToday;
};
