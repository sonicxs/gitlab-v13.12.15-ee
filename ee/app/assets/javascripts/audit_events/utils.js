import { parsePikadayDate, pikadayToString } from '~/lib/utils/datetime_utility';
import { AVAILABLE_TOKEN_TYPES, AUDIT_FILTER_CONFIGS } from './constants';

export const getTypeFromEntityType = (entityType) => {
  return AUDIT_FILTER_CONFIGS.find(
    ({ entityType: configEntityType }) => configEntityType === entityType,
  )?.type;
};

export const getEntityTypeFromType = (type) => {
  return AUDIT_FILTER_CONFIGS.find(({ type: configType }) => configType === type)?.entityType;
};

export const parseAuditEventSearchQuery = ({
  created_after: createdAfter,
  created_before: createdBefore,
  entity_type: entityType,
  ...restOfParams
}) => ({
  ...restOfParams,
  created_after: createdAfter ? parsePikadayDate(createdAfter) : null,
  created_before: createdBefore ? parsePikadayDate(createdBefore) : null,
  entity_type: getTypeFromEntityType(entityType),
});

export const createAuditEventSearchQuery = ({ filterValue, startDate, endDate, sortBy }) => {
  const entityValue = filterValue.find((value) => AVAILABLE_TOKEN_TYPES.includes(value.type));

  return {
    created_after: startDate ? pikadayToString(startDate) : null,
    created_before: endDate ? pikadayToString(endDate) : null,
    sort: sortBy,
    entity_id: entityValue?.value.data,
    entity_type: getEntityTypeFromType(entityValue?.type),
    // When changing the search parameters, we should be resetting to the first page
    page: null,
  };
};
