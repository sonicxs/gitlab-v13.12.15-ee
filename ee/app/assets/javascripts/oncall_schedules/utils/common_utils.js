import { newDateAsLocaleTime } from '~/lib/utils/datetime_utility';
import { sprintf, __ } from '~/locale';
import { ASSIGNEE_COLORS_COMBO } from '../constants';

/**
 * Returns formatted timezone string, e.g. (UTC-09:00) AKST Alaska
 *
 * @param {Object} tz
 * @param {String} tz.name
 * @param {String} tz.formatted_offset
 * @param {String} tz.abbr
 *
 * @returns {String}
 */
export const getFormattedTimezone = (tz) => {
  return sprintf(__('(UTC %{offset}) %{timezone}'), {
    offset: tz.formatted_offset,
    timezone: `${tz.abbr} ${tz.name}`,
  });
};

/**
 * Returns `true` for non-empty string, otherwise returns `false`
 *
 * @param {String} startDate
 *
 * @returns {Boolean}
 */
export const isNameFieldValid = (nameField) => {
  return Boolean(nameField?.length);
};

/**
 * Returns a Array of Objects that represent the shift participant
 * with his/her username and unique shift color values
 *
 * @param {Object[]} participants
 * @param {string} participants[].username - The username of the participant.
 *
 * @returns {Object[]} A list of values to save each participant
 * @property {string} username
 * @property {string} colorWeight
 * @property {string} colorPalette
 */
export const getParticipantsForSave = (participants) =>
  participants.map(({ username }, index) => {
    const colorIndex = index % ASSIGNEE_COLORS_COMBO.length;
    const { colorWeight, colorPalette } = ASSIGNEE_COLORS_COMBO[colorIndex];

    return {
      username,
      colorWeight,
      colorPalette,
    };
  });

/**
 * Parses a activePeriod string into an integer value
 *
 * @param {String} hourString
 */
export const parseHour = (hourString) => parseInt(hourString.slice(0, 2), 10);

/**
 * Parses a rotation date for use in the add/edit rotation form
 *
 * @param {ISOString} dateTimeString
 * @param {Timezone string - long} scheduleTimezone
 */
export const parseRotationDate = (dateTimeString, scheduleTimezone) => {
  const options = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    hourCycle: 'h23',
    timeZone: scheduleTimezone,
    timeZoneName: 'long',
  };
  const formatter = new Intl.DateTimeFormat('en-US', options);
  const parts = formatter.formatToParts(Date.parse(dateTimeString));
  const [month, , day, , year, , hour] = parts.map((part) => part.value);
  // The datepicker uses local time
  const date = newDateAsLocaleTime(`${year}-${month}-${day}`);
  const time = parseInt(hour, 10);

  return { date, time };
};
