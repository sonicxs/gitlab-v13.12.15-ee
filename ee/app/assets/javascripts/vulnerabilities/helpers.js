import { isAbsolute, isSafeURL } from '~/lib/utils/url_utility';
import { REGEXES, gidPrefix, uidPrefix } from './constants';

// Get the issue in the format expected by the descendant components of related_issues_block.vue.
export const getFormattedIssue = (issue) => ({
  ...issue,
  reference: `#${issue.iid}`,
  path: issue.web_url,
});

export const getAddRelatedIssueRequestParams = (reference, defaultProjectId) => {
  let issueId = reference;
  let projectId = defaultProjectId;

  // If the reference is an issue number, parse out just the issue number.
  if (REGEXES.ISSUE_FORMAT.test(reference)) {
    [, issueId] = REGEXES.ISSUE_FORMAT.exec(reference);
  }
  // If the reference is an absolute URL and matches the issues URL format, parse out the project and issue.
  else if (isSafeURL(reference) && isAbsolute(reference)) {
    const { pathname } = new URL(reference);

    if (REGEXES.LINK_FORMAT.test(pathname)) {
      [, projectId, issueId] = REGEXES.LINK_FORMAT.exec(pathname);
    }
  }

  return { target_issue_iid: issueId, target_project_id: projectId };
};

export const normalizeGraphQLVulnerability = (vulnerability) => {
  if (!vulnerability) {
    return null;
  }

  const newVulnerability = { ...vulnerability };

  if (vulnerability.id) {
    newVulnerability.id = vulnerability.id.replace(gidPrefix, '');
  }

  if (vulnerability.state) {
    newVulnerability.state = vulnerability.state.toLowerCase();
  }

  ['confirmed', 'resolved', 'dismissed'].forEach((state) => {
    if (vulnerability[`${state}By`]?.id) {
      newVulnerability[`${state}ById`] = vulnerability[`${state}By`].id.replace(uidPrefix, '');
      delete newVulnerability[`${state}By`];
    }
  });

  return newVulnerability;
};
