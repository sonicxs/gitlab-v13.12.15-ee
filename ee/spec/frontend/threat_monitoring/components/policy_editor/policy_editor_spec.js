import { GlModal, GlToggle } from '@gitlab/ui';
import { shallowMount } from '@vue/test-utils';
import {
  RuleDirectionInbound,
  PortMatchModeAny,
  RuleTypeEndpoint,
  EditorModeYAML,
  EndpointMatchModeLabel,
} from 'ee/threat_monitoring/components/policy_editor/constants';
import fromYaml from 'ee/threat_monitoring/components/policy_editor/lib/from_yaml';
import { buildRule } from 'ee/threat_monitoring/components/policy_editor/lib/rules';
import toYaml from 'ee/threat_monitoring/components/policy_editor/lib/to_yaml';
import PolicyAlertPicker from 'ee/threat_monitoring/components/policy_editor/policy_alert_picker.vue';
import PolicyEditorApp from 'ee/threat_monitoring/components/policy_editor/policy_editor.vue';
import PolicyPreview from 'ee/threat_monitoring/components/policy_editor/policy_preview.vue';
import PolicyRuleBuilder from 'ee/threat_monitoring/components/policy_editor/policy_rule_builder.vue';
import createStore from 'ee/threat_monitoring/store';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';
import { redirectTo } from '~/lib/utils/url_utility';

jest.mock('~/lib/utils/url_utility');

describe('PolicyEditorApp component', () => {
  let store;
  let wrapper;
  const l7manifest = `apiVersion: cilium.io/v2
  kind: CiliumNetworkPolicy
metadata:
  name: limit-inbound-ip
spec:
  endpointSelector: {}
  ingress:
  - toPorts:
    - ports:
      - port: '80'
        protocol: TCP
      - port: '443'
        protocol: TCP
      rules:
        http:
        - headers:
          - 'X-Forwarded-For: 192.168.1.1'
    fromEntities:
    - cluster`;

  const factory = ({ propsData, provide = {}, state, data } = {}) => {
    store = createStore();
    Object.assign(store.state.threatMonitoring, {
      ...state,
    });
    Object.assign(store.state.networkPolicies, {
      ...state,
    });

    jest.spyOn(store, 'dispatch').mockImplementation(() => Promise.resolve());

    wrapper = extendedWrapper(
      shallowMount(PolicyEditorApp, {
        propsData: {
          threatMonitoringPath: '/threat-monitoring',
          projectId: '21',
          ...propsData,
        },
        provide: {
          glFeatures: { threatMonitoringAlerts: false },
          ...provide,
        },
        store,
        data,
        stubs: { NetworkPolicyEditor: true },
      }),
    );
  };

  const findRuleEditor = () => wrapper.findByTestId('rule-editor');
  const findYamlEditor = () => wrapper.findByTestId('yaml-editor');
  const findPreview = () => wrapper.find(PolicyPreview);
  const findAddRuleButton = () => wrapper.findByTestId('add-rule');
  const findYAMLParsingAlert = () => wrapper.findByTestId('parsing-alert');
  const findNetworkPolicyEditor = () => wrapper.findByTestId('network-policy-editor');
  const findPolicyAlertPicker = () => wrapper.find(PolicyAlertPicker);
  const findPolicyDescription = () => wrapper.find("[id='policyDescription']");
  const findPolicyEnableContainer = () => wrapper.findByTestId('policy-enable');
  const findPolicyName = () => wrapper.find("[id='policyName']");
  const findPolicyRuleBuilder = () => wrapper.findComponent(PolicyRuleBuilder);
  const findSavePolicy = () => wrapper.findByTestId('save-policy');
  const findDeletePolicy = () => wrapper.findByTestId('delete-policy');
  const findEditorModeToggle = () => wrapper.findByTestId('editor-mode');

  const modifyPolicyAlert = async ({ isAlertEnabled }) => {
    const policyAlertPicker = findPolicyAlertPicker();
    policyAlertPicker.vm.$emit('update-alert', isAlertEnabled);
    await wrapper.vm.$nextTick();
    expect(policyAlertPicker.props('policyAlert')).toBe(isAlertEnabled);
    findSavePolicy().vm.$emit('click');
    await wrapper.vm.$nextTick();
  };

  beforeEach(() => {
    factory();
  });

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  it('renders the policy editor layout', () => {
    expect(wrapper.find('section').element).toMatchSnapshot();
  });

  it('renders toggle with label', () => {
    expect(wrapper.findComponent(GlToggle).props('label')).toBe(PolicyEditorApp.i18n.toggleLabel);
  });

  it('renders a default rule with label', () => {
    expect(findPolicyRuleBuilder().exists()).toBe(true);
    expect(findPolicyRuleBuilder().attributes()).toMatchObject({
      endpointlabels: '',
      endpointmatchmode: 'any',
    });
  });

  it('does not render yaml editor', () => {
    expect(findYamlEditor().exists()).toBe(false);
  });

  it('does not render parsing error alert', () => {
    expect(findYAMLParsingAlert().exists()).toBe(false);
  });

  it('does not render delete button', () => {
    expect(findDeletePolicy().exists()).toBe(false);
  });

  it('does not render the policy alert picker', () => {
    expect(findPolicyAlertPicker().exists()).toBe(false);
  });

  describe('given .yaml editor mode is enabled', () => {
    beforeEach(() => {
      factory({
        data: () => ({
          editorMode: EditorModeYAML,
        }),
      });
    });

    it('does not render rule editor', () => {
      expect(findRuleEditor().exists()).toBe(false);
    });

    it('renders yaml editor', () => {
      const editor = findYamlEditor();
      expect(editor.exists()).toBe(true);
      expect(editor.element).toMatchSnapshot();
    });

    it('updates policy on yaml editor value change', async () => {
      const manifest = `apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
description: test description
metadata:
  name: test-policy
  labels:
    app.gitlab.com/proj: '21'
spec:
  endpointSelector:
    matchLabels:
      network-policy.gitlab.com/disabled_by: gitlab
      foo: bar
  ingress:
  - fromEndpoints:
    - matchLabels:
        foo: bar`;
      findNetworkPolicyEditor().vm.$emit('input', manifest);

      expect(wrapper.vm.policy).toMatchObject({
        name: 'test-policy',
        description: 'test description',
        isEnabled: false,
        endpointMatchMode: EndpointMatchModeLabel,
        endpointLabels: 'foo:bar',
        rules: [
          {
            ruleType: RuleTypeEndpoint,
            matchLabels: 'foo:bar',
          },
        ],
        labels: { 'app.gitlab.com/proj': '21' },
      });
    });

    it('saves L7 policies', async () => {
      factory({
        data: () => ({
          editorMode: EditorModeYAML,
          yamlEditorValue: l7manifest,
        }),
      });
      findSavePolicy().vm.$emit('click');

      await wrapper.vm.$nextTick();
      expect(store.dispatch).toHaveBeenCalledWith('networkPolicies/createPolicy', {
        environmentId: -1,
        policy: { manifest: l7manifest },
      });
      expect(redirectTo).toHaveBeenCalledWith('/threat-monitoring');
    });
  });

  describe('given there is a name change', () => {
    let initialValue;

    beforeEach(() => {
      initialValue = findPreview().props('policyYaml');
      findPolicyName().vm.$emit('input', 'new');
    });

    it('updates policy yaml preview', () => {
      expect(findPreview().props('policyYaml')).not.toEqual(initialValue);
    });
  });

  describe('given there is a rule change', () => {
    let initialValue;

    beforeEach(() => {
      initialValue = findPreview().props('policyDescription');
      wrapper.findByTestId('add-rule').vm.$emit('click');
    });

    it('updates policy description preview', () => {
      expect(findPreview().props('policyDescription')).not.toEqual(initialValue);
    });
  });

  it('adds a new rule', async () => {
    expect(wrapper.findAllComponents(PolicyRuleBuilder)).toHaveLength(1);
    const button = findAddRuleButton();
    button.vm.$emit('click');
    button.vm.$emit('click');
    await wrapper.vm.$nextTick();
    const elements = wrapper.findAllComponents(PolicyRuleBuilder);
    expect(elements).toHaveLength(3);

    elements.wrappers.forEach((builder, idx) => {
      expect(builder.props().rule).toMatchObject({
        ruleType: RuleTypeEndpoint,
        direction: RuleDirectionInbound,
        matchLabels: '',
        portMatchMode: PortMatchModeAny,
        ports: '',
      });
      expect(builder.props().endpointSelectorDisabled).toEqual(idx !== 0);
    });
  });

  it('removes a new rule', async () => {
    findAddRuleButton().vm.$emit('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.findAllComponents(PolicyRuleBuilder)).toHaveLength(2);

    findPolicyRuleBuilder().vm.$emit('remove');
    await wrapper.vm.$nextTick();
    expect(wrapper.findAllComponents(PolicyRuleBuilder)).toHaveLength(1);
  });

  it('updates yaml editor value on switch to yaml editor', async () => {
    findPolicyName().vm.$emit('input', 'test-policy');
    findEditorModeToggle().vm.$emit('input', EditorModeYAML);
    await wrapper.vm.$nextTick();

    const editor = findNetworkPolicyEditor();
    expect(editor.exists()).toBe(true);
    expect(fromYaml(editor.attributes('value'))).toMatchObject({
      name: 'test-policy',
    });
  });

  describe('given there is a yaml parsing error', () => {
    beforeEach(() => {
      factory({
        data: () => ({
          yamlEditorError: {},
        }),
      });
    });

    it('disables policy name field', () => {
      expect(findPolicyName().attributes().disabled).toBe('true');
    });

    it('disables policy description field', () => {
      expect(findPolicyDescription().attributes().disabled).toBe('true');
    });

    it('disables policy enable/disable toggle', () => {
      expect(findPolicyEnableContainer().attributes().disabled).toBe('true');
    });

    it('renders parsing error alert', () => {
      expect(findYAMLParsingAlert().exists()).toBe(true);
    });

    it('disables rule builder', () => {
      expect(wrapper.findByTestId('rule-builder-container').props().disabled).toBe(true);
    });

    it('disables action picker', () => {
      expect(wrapper.findByTestId('policy-action-container').props().disabled).toBe(true);
    });

    it('disables policy preview', () => {
      expect(wrapper.findByTestId('policy-preview-container').props().disabled).toBe(true);
    });

    it('does not update yaml editor value on switch to yaml editor', async () => {
      findPolicyName().vm.$emit('input', 'test-policy');
      findEditorModeToggle().vm.$emit('input', EditorModeYAML);
      await wrapper.vm.$nextTick();

      const editor = findNetworkPolicyEditor();
      expect(editor.exists()).toBe(true);
      expect(editor.attributes('value')).toEqual('');
    });
  });

  it('creates policy and redirects to a threat monitoring path', async () => {
    findSavePolicy().vm.$emit('click');

    await wrapper.vm.$nextTick();
    expect(store.dispatch).toHaveBeenCalledWith('networkPolicies/createPolicy', {
      environmentId: -1,
      policy: { manifest: toYaml(wrapper.vm.policy) },
    });
    expect(redirectTo).toHaveBeenCalledWith('/threat-monitoring');
  });

  describe('given there is a createPolicy error', () => {
    beforeEach(() => {
      factory({
        state: {
          errorUpdatingPolicy: true,
        },
      });
    });

    it('it does not redirect', async () => {
      findSavePolicy().vm.$emit('click');

      await wrapper.vm.$nextTick();
      expect(redirectTo).not.toHaveBeenCalledWith('/threat-monitoring');
    });
  });

  describe('given existingPolicy property was provided', () => {
    const manifest = toYaml({
      name: 'policy',
      endpointLabels: '',
      rules: [buildRule()],
    });

    beforeEach(() => {
      factory({
        propsData: {
          existingPolicy: { name: 'policy', manifest },
        },
      });
    });

    it('presents existing policy', () => {
      expect(findPolicyName().attributes().value).toEqual('policy');
      expect(wrapper.findAllComponents(PolicyRuleBuilder)).toHaveLength(1);
    });

    it('updates existing policy and redirects to a threat monitoring path', async () => {
      const saveButton = findSavePolicy();
      expect(saveButton.text()).toEqual('Save changes');
      saveButton.vm.$emit('click');

      await wrapper.vm.$nextTick();
      expect(store.dispatch).toHaveBeenCalledWith('networkPolicies/updatePolicy', {
        environmentId: -1,
        policy: { name: 'policy', manifest: toYaml(wrapper.vm.policy) },
      });
      expect(redirectTo).toHaveBeenCalledWith('/threat-monitoring');
    });

    describe('given there is a updatePolicy error', () => {
      beforeEach(() => {
        factory({
          propsData: {
            existingPolicy: { name: 'policy', manifest },
          },
          state: {
            errorUpdatingPolicy: true,
          },
        });
      });

      it('it does not redirect', async () => {
        findSavePolicy().vm.$emit('click');

        await wrapper.vm.$nextTick();
        expect(redirectTo).not.toHaveBeenCalledWith('/threat-monitoring');
      });
    });

    it('renders delete button', () => {
      expect(findDeletePolicy().exists()).toBe(true);
    });

    it('it does not trigger deletePolicy on delete button click', async () => {
      findDeletePolicy().vm.$emit('click');
      await wrapper.vm.$nextTick();

      expect(store.dispatch).not.toHaveBeenCalledWith('networkPolicies/deletePolicy');
    });

    it('removes policy and redirects to a threat monitoring path on secondary modal button click', async () => {
      wrapper.find(GlModal).vm.$emit('secondary');
      await wrapper.vm.$nextTick();

      expect(store.dispatch).toHaveBeenCalledWith('networkPolicies/deletePolicy', {
        environmentId: -1,
        policy: { name: 'policy', manifest: toYaml(wrapper.vm.policy) },
      });
      expect(redirectTo).toHaveBeenCalledWith('/threat-monitoring');
    });
  });

  describe('add alert picker', () => {
    beforeEach(() => {
      factory({ provide: { glFeatures: { threatMonitoringAlerts: true } } });
    });

    it('does render the policy alert picker', () => {
      expect(findPolicyAlertPicker().exists()).toBe(true);
    });

    it('adds a policy annotation on alert addition', async () => {
      await modifyPolicyAlert({ isAlertEnabled: true });
      expect(store.dispatch).toHaveBeenLastCalledWith('networkPolicies/createPolicy', {
        environmentId: -1,
        policy: {
          manifest: expect.stringContaining("app.gitlab.com/alert: 'true'"),
        },
      });
    });

    it('removes a policy annotation on alert removal', async () => {
      await modifyPolicyAlert({ isAlertEnabled: false });
      expect(store.dispatch).toHaveBeenLastCalledWith('networkPolicies/createPolicy', {
        environmentId: -1,
        policy: {
          manifest: expect.not.stringContaining("app.gitlab.com/alert: 'true'"),
        },
      });
    });
  });
});
