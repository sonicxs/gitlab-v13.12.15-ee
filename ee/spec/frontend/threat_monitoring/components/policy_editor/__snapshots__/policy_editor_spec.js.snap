// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyEditorApp component given .yaml editor mode is enabled renders yaml editor 1`] = `
<div
  class="row"
  data-testid="yaml-editor"
>
  <div
    class="col-sm-12 col-md-12 col-lg-10 col-xl-8"
  >
    <div
      class="gl-rounded-base gl-border-1 gl-border-solid gl-border-gray-100"
    >
      <h5
        class="gl-m-0 gl-p-4 gl-bg-gray-10 gl-border-1 gl-border-b-solid gl-border-b-gray-100"
      >
        
          YAML editor
        
      </h5>
       
      <div
        class="gl-p-4"
      >
        <networkpolicyeditor-stub
          data-testid="network-policy-editor"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PolicyEditorApp component renders the policy editor layout 1`] = `
<section>
  <header
    class="my-3"
  >
    <h2
      class="h3 mb-1"
    >
      
      Policy description
    
    </h2>
  </header>
   
  <div
    class="row"
  >
    <div
      class="col-sm-6 col-md-4 col-lg-3 col-xl-2"
    >
      <gl-form-group-stub
        label="Policy type"
        label-for="policyType"
      >
        <gl-form-select-stub
          disabled=""
          id="policyType"
          options="[object Object]"
          value="networkPolicy"
        />
      </gl-form-group-stub>
    </div>
     
    <div
      class="col-sm-6 col-md-6 col-lg-5 col-xl-4"
    >
      <gl-form-group-stub
        label="Name"
        label-for="policyName"
      >
        <gl-form-input-stub
          id="policyName"
          value=""
        />
      </gl-form-group-stub>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-sm-12 col-md-10 col-lg-8 col-xl-6"
    >
      <gl-form-group-stub
        label="Description"
        label-for="policyDescription"
      >
        <gl-form-textarea-stub
          id="policyDescription"
          noresize="true"
          value=""
        />
      </gl-form-group-stub>
    </div>
  </div>
   
  <div
    class="row"
  >
    <environment-picker-stub />
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-md-auto"
    >
      <gl-form-group-stub
        data-testid="policy-enable"
      >
        <gl-toggle-stub
          label="Policy status"
          labelposition="top"
        />
      </gl-form-group-stub>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-md-auto"
    >
      <gl-form-group-stub
        label="Editor mode"
        label-for="editorMode"
      >
        <gl-segmented-control-stub
          checked="rule"
          data-testid="editor-mode"
          options="[object Object],[object Object]"
        />
      </gl-form-group-stub>
    </div>
  </div>
   
  <hr />
   
  <div
    class="row"
    data-testid="rule-editor"
  >
    <div
      class="col-sm-12 col-md-6 col-lg-7 col-xl-8"
    >
      <!---->
       
      <dim-disable-container-stub
        data-testid="rule-builder-container"
      >
          
        <policy-rule-builder-stub
          class="gl-mb-4"
          endpointlabels=""
          endpointmatchmode="any"
          rule="[object Object]"
        />
         
        <div
          class="gl-p-3 gl-rounded-base gl-border-1 gl-border-solid gl-border-gray-100 gl-mb-5"
        >
          <gl-button-stub
            buttontextclasses=""
            category="primary"
            data-testid="add-rule"
            icon=""
            size="medium"
            variant="link"
          >
            New rule
          </gl-button-stub>
        </div>
      </dim-disable-container-stub>
       
      <dim-disable-container-stub
        data-testid="policy-action-container"
      >
          
        <policy-action-picker-stub />
         
        <!---->
      </dim-disable-container-stub>
    </div>
     
    <div
      class="col-sm-12 col-md-6 col-lg-5 col-xl-4"
    >
      <dim-disable-container-stub
        data-testid="policy-preview-container"
      >
          
        <policy-preview-stub
          initialtab="0"
          policydescription="Allow all inbound traffic to <strong>all</strong> pods from <strong>all</strong> pods on <strong>any</strong> port"
          policyyaml="apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: ''
  labels:
    app.gitlab.com/proj: '21'
spec:
  endpointSelector:
    matchLabels:
      network-policy.gitlab.com/disabled_by: gitlab
  ingress:
  - fromEndpoints:
    - matchLabels: {}
"
        />
      </dim-disable-container-stub>
    </div>
  </div>
   
  <!---->
   
  <hr />
   
  <div
    class="row"
  >
    <div
      class="col-md-auto"
    >
      <gl-button-stub
        buttontextclasses=""
        category="primary"
        data-testid="save-policy"
        icon=""
        size="medium"
        type="submit"
        variant="success"
      >
        Create policy
      </gl-button-stub>
       
      <!---->
       
      <gl-button-stub
        buttontextclasses=""
        category="secondary"
        href="/threat-monitoring"
        icon=""
        size="medium"
        variant="default"
      >
        Cancel
      </gl-button-stub>
    </div>
  </div>
   
  <gl-modal-stub
    actioncancel="[object Object]"
    actionsecondary="[object Object]"
    dismisslabel="Close"
    modalclass=""
    modalid="delete-modal"
    size="md"
    title="Delete policy: "
    titletag="h4"
  >
    
    Are you sure you want to delete this policy? This action cannot be undone.
  
  </gl-modal-stub>
</section>
`;
